tutorials/haskell/hspec-golden-test-tutorial/tutorial.md

```diff
@@ -0,0 +1,506 @@
+
+
+## A More Real Test Case
+
+The `fizzbuzz` module was used to create a simple and didactic example but it could have been easily tested with unit tests. In the practice golden tests are often used when testing JSON, HTML or images which generate large outputs. We haven't also tested the whole functionality of `hspec-golden` library yet. For example the `hgold` CLI for updating the golden files.
+
+Lets create some new modules to demostrate a real case in which golden tests are needed.
+
+- Html.hs : Renders a HTML template.
+- Json.hs : Encodes a data type into a JSON Bytestring.  
+
+For convenience each module will be placed inside its own directory.
+
+```shell
+$ tree src
+src
+├── FIZZBUZZ
+│   └── FizzBuzz.hs
+├── HelloWorld.hs
+├── HTML
+│   └── Html.hs
+└── JSON
+    └── Json.hs
+
+```
+
+To make things easy and save time, let's add the following code into each module.
+
+###### Html.hs
+
+```haskell
+{-# LANGUAGE OverloadedStrings #-}
+
+
+module HTML.Html  where
+
+import           Text.Blaze.Html5               as H
+import           Text.Blaze.Html5.Attributes    as A
+import           Text.Blaze.Html.Renderer.Pretty        (renderHtml)
+
+
+htmlRendered :: Html -> String
+htmlRendered page = renderHtml page
+
+somePage :: Html
+somePage = html $ do
+    H.head $ do
+        H.title "StackBuilders Tutorial."
+    body $ do
+       "Hello Golden Testers."
+```
+
+
+###### Json.hs
+
+```haskell
+{-# LANGUAGE DeriveGeneric #-}
+
+module JSON.Json where
+
+import            Data.Aeson                   (ToJSON, encode)
+import            GHC.Generics                 (Generic)
+import            Data.ByteString.Lazy         (ByteString)
+
+data Country = Country
+  {
+    cname      :: String,
+    continent :: String,
+    ctag       :: Int
+  } deriving (Generic, Show)
+
+instance ToJSON Country
+
+
+ecuador = Country "Ecuador" "America" 1
+germany = Country "Germany" "Europe" 2
+japan = Country "Japan" "Asia" 3
+
+countries :: [Country]
+countries = [ecuador,germany,japan]
+
+encodeCountries :: [Country] -> ByteString
+encodeCountries = encode
+```
+
+Before coding our tests, lets create and organize `HtmlGoldenSpec.hs` and `JsonGoldenSpec.hs` modules into directories as well.
+
+```shell
+$ tree tests
+test
+├── FizzBuzz
+│   └── FizzBuzzGoldenSpec.hs
+├── Hello
+│   ├── HelloGoldenSpec.hs
+│   └── HelloSpec.hs
+├── Html
+│   └── HtmlGoldenSpec.hs
+├── Json
+│   └── JsonGoldenSpec.hs
+└── Spec.hs
+```
+We are now ready to start coding our golden tests.
+
+#### HTML Golden tests
+
+Lets start with the HTML module importing all the things we need
```

Comment:
@Jagl257 We probably want to add a colon (`:`) that introduces the following lines.

Reply:
@CristhianMotoche the colon (:) is used for enumerations only and in this case I'm just showing a code block. Anyway I think a dot should work.